indra/llconvexdecomposition/llconvexdecompositionvhacd.cpp

```diff
@@ -0,0 +1,492 @@
+    mDecompParams.push_back(param);
+
+    param.mName = "Num Vertices";
+    param.mDescription = nullptr;
+    param.mType = LLCDParam::LLCD_FLOAT;
+    param.mDetails.mRange.mLow.mFloat = 3.f;
+    param.mDetails.mRange.mHigh.mFloat = MAX_VERTICES_PER_HULL;
+    param.mDetails.mRange.mDelta.mFloat = 1.f;
+    param.mDefault.mFloat = 32.f;
+    param.mStage = 0;
+    param.mReserved = -1;
+    mDecompParams.push_back(param);
+
+    param.mName = "Error Tolerance";
+    param.mDescription = nullptr;
+    param.mType = LLCDParam::LLCD_FLOAT;
+    param.mDetails.mRange.mLow.mFloat = 0.0001f;
+    param.mDetails.mRange.mHigh.mFloat = 99.f;
+    param.mDetails.mRange.mDelta.mFloat = 0.001f;
+    param.mDefault.mFloat = 1.f;
+    param.mStage = 0;
+    param.mReserved = -1;
+    mDecompParams.push_back(param);
+
+    for (const LLCDParam& param : mDecompParams)
+    {
+        const char* const name = param.mName;
+
+        switch (param.mType)
+        {
+        case LLCDParam::LLCD_FLOAT:
+        {
+            setParam(name, param.mDefault.mFloat);
+            break;
+        }
+        case LLCDParam::LLCD_ENUM:
+        case LLCDParam::LLCD_INTEGER:
+        {
+            setParam(name, param.mDefault.mIntOrEnumValue);
+            break;
+        }
+        case LLCDParam::LLCD_BOOLEAN:
+        {
+            setParam(name, (param.mDefault.mBool != 0));
+            break;
+        }
+        case LLCDParam::LLCD_INVALID:
+        default:
+        {
+            break;
+        }
+        }
+    }
+}
+
+LLConvexDecompositionVHACD::~LLConvexDecompositionVHACD()
+{
+    mBoundDecomp = nullptr;
+    mDecompData.clear();
+
+    mVHACD->Release();
+}
+
+void LLConvexDecompositionVHACD::genDecomposition(int& decomp)
+{
+    int new_decomp_id = static_cast<int>(mDecompData.size()) + 1;
+    mDecompData[new_decomp_id] = LLDecompData();
+    decomp = new_decomp_id;
+}
+
+void LLConvexDecompositionVHACD::deleteDecomposition(int decomp)
+{
+    auto iter = mDecompData.find(decomp);
+    if (iter != mDecompData.end())
+    {
+        if (mBoundDecomp == &iter->second)
+        {
+            mBoundDecomp = nullptr;
+        }
+        mDecompData.erase(iter);
+    }
+}
+
+void LLConvexDecompositionVHACD::bindDecomposition(int decomp)
+{
+    auto iter = mDecompData.find(decomp);
+    if (iter != mDecompData.end())
+    {
+        mBoundDecomp = &iter->second;
+    }
+    else
+    {
+        LL_WARNS() << "Failed to bind unknown decomposition: " << decomp << LL_ENDL;
+        mBoundDecomp = nullptr;
+    }
+}
+
+LLCDResult LLConvexDecompositionVHACD::setParam(const char* name, float val)
+{
+    if (name == std::string("Num Hulls"))
+    {
+        mVHACDParameters.m_maxConvexHulls = llclamp(ll_round(val), 1, MAX_HULLS);
+    }
+    else if (name == std::string("Num Vertices"))
+    {
+        mVHACDParameters.m_maxNumVerticesPerCH = llclamp(ll_round(val), 3, MAX_VERTICES_PER_HULL);
+    }
+    else if (name == std::string("Error Tolerance"))
```

Comment:
Avoid string temporaries
```suggestion
    if (strcmp(name, "Num Hulls") == 0)
    {
        mVHACDParameters.m_maxConvexHulls = llclamp(ll_round(val), 1, MAX_HULLS);
    }
    else if (strcmp(name, "Num Vertices") == 0)
    {
        mVHACDParameters.m_maxNumVerticesPerCH = llclamp(ll_round(val), 3, MAX_VERTICES_PER_HULL);
    }
    else if (strcmp(name, "Error Tolerance") == 0)
```
Or if you have string_view
```suggestion
    if (name == std::string_view("Num Hulls"))
    {
        mVHACDParameters.m_maxConvexHulls = llclamp(ll_round(val), 1, MAX_HULLS);
    }
    else if (name == std::string_view("Num Vertices"))
    {
        mVHACDParameters.m_maxNumVerticesPerCH = llclamp(ll_round(val), 3, MAX_VERTICES_PER_HULL);
    }
    else if (name == std::string_view("Error Tolerance"))
```